### DISCUSSION TITLE
How to execute CPD?

### DISCUSSION THREAD

**Discussion by T-chuangxin** [pmd/pmd#3036] [Q&A] [answered]
any cpd demos? I want to run cpd in my plugin project . Tks!!

**Comment by oowekyala** [comment]
Hello, CLI usage is explained there 

https://pmd.github.io/pmd-6.6.0/pmd_userdocs_installation.html#running-cpd-via-command-line

Don't forget to set the `--language` option (you didn't specify any). More info on CPD CLI usage and Ant task usage is available here

https://pmd.github.io/pmd-6.6.0/pmd_userdocs_cpd.html

If you'd like to integrate it in a maven build, see

https://pmd.github.io/pmd-6.6.0/pmd_userdocs_tools_maven.html#integrated-into-the-build-process

which mentions using `cpd-check` as a goal.

**Comment by T-chuangxin** [comment]
I means use the CPD API to run ,my project has add the jar files!

**Comment by jsotuyod** [comment, marked as answer]
@T-chuangxin if you want to run CPD programatically, I suggest you have a look at how PMD does it itself

https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDCommandLineInterface.java#L59

**Comment by T-chuangxin** [comment]
@jsotuyod Aha, thank you for your reply, the function has been made! Here is the code:

 val project = e.project ?: return
        val cpdConfiguration = CPDConfiguration()
        //token
        cpdConfiguration.minimumTileSize = 100
        //java
        cpdConfiguration.language = JavaLanguage()
        cpdConfiguration.encoding = "UTF-8"
        //Print help text
        cpdConfiguration.isHelp=true
        //project
        cpdConfiguration.files=traverseFileTree(project.basePath.toString())
        cpdConfiguration.renderer=XMLRenderer()
        val cpd=CPD(cpdConfiguration)
        tokenizeFiles(cpd,traverseFileTree(project.basePath.toString()))
        cpd.go()
        report(cpd,project.basePath.toString())
